### PR TITLE
feat: adding simple models and solvers

### PR DESCRIPTION
**Pull request by ConnorStoneAstro**
This PR integrates the simple models (GRF, MVG, joint, convolved_likelihood, etc.) into score_models. It also integrates dedicated SDE and ODE solvers. For now the solvers are only integrated for the `score_model.sample` function. This is a draft PR to discuss API details and further integration.

### CODE REVIEW

File: score_models/sbm/score_model.py [hunk 1]
```diff
@@ -66,76 +63,92 @@ def ode_drift(self, t, x, *args) -> Tensor:
...
         """
-        Compute the log likelihood of point x using the probability flow ODE, 
+        Compute the log likelihood of point x using the probability flow ODE,
         which makes use of the instantaneous change of variable formula
         developed by Chen et al. 2018 (arxiv.org/abs/1806.07366).
         See Song et al. 2020 (arxiv.org/abs/2011.13456) for usage with SDE formalism of SBM.
         """
-        drift = self.ode_drift
-        hessian_trace = lambda t, x, *args: self.hessian_trace(t, x, *args, **kwargs)
+        if method == "Euler_ODE":
```
Comment (AlexandreAdam) [inline score_models/sbm/score_model.py]: method should be lowercase names, and made simple to use using the Literal type. 

e.g.
```
def log_likelihood(..., method: Literal["euler", "heun", "rk2", "rk4"] = "euler")
```

And we can allow capital letter by using the lower method on the string
```
if method.lower() == "euler":
...
```

File: score_models/solver/ode.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+    def forward(self, x0, N, **kwargs):
+        """Call this to solve the ODE forward in time from x0 at time t_min to xT at time t_max"""
+        return self._solve(x0, N, self.forward_dx, True, **kwargs)
+
+    def reverse(self, xT, N, **kwargs):
+        """Call this to solve the ODE backward in time from xT at time t_max to x0 at time t_min"""
+        return self._solve(xT, N, self.reverse_dx, False, **kwargs)
+
+    def forward_dx(self, t, x, dt):
```
Comment (AlexandreAdam) [inline score_models/solver/ode.py]: The probability flow ODE has the same drift forward and backward in time. The drift you wrote for the reverse_dx is the correct one.

File: score_models/solver/ode.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+        """Forward discretization of the ODE, this is the update for x"""
+        return self.sde.drift(t, x) * dt
+
+    def reverse_dx(self, t, x, dt, **kwargs):
+        """Reverse discretization of the ODE, this is the update for x"""
+        return (
```
Comment (AlexandreAdam) [inline score_models/solver/ode.py]: I like to write this in the following way
```
f = self.sde.drift(t, x)
g = self.sde.diffusion(t, x)
s = self.score(t, x, *args) # args are for conditional arguments
return f - 0.5 * g**2 * s
```

**Review by AlexandreAdam — COMMENTED** [review]
Main thing is the PFODE drift situation. The drift is the same for the forward and reverse-time case.

File: score_models/analytic_models/annealed.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+import torch
+import torch.nn as nn
+from torch import Tensor
+
+from ..sde import SDE
+from ..sbm import ScoreModel
+
+
+class AnnealedScoreModel(nn.Module):
```
Comment (AlexandreAdam) [inline score_models/analytic_models/annealed.py]: I would choose a different name for this to reflect its usage better. You are stitching two score models together. Maybe something along the lines of InterpolateScoreModels

File: score_models/analytic_models/conv_likelihood.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+    """
+
+    @torch.no_grad()
+    def __init__(
+        self,
+        sde: SDE,
+        y: Tensor,
+        Sigma_y: Tensor,
+        x_shape: Tuple[int],
+        A: Optional[Tensor] = None,
```
Comment (AlexandreAdam) [inline score_models/analytic_models/conv_likelihood.py]: It is common for this approximation to use the functional representation of the forward model. So either add a way to define a forward model or A should accept callable types

File: score_models/analytic_models/grf.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
+import torch
+import torch.nn as nn
+from torch import Tensor
+
+from ..sde import SDE
+
+
+class GRFEnergyModel(nn.Module):
```
Comment (AlexandreAdam) [inline score_models/analytic_models/grf.py]: Can this inherit from the Energy model class?

File: score_models/analytic_models/conv_likelihood.py (hunk 5 above)
Comment (ConnorStoneAstro) [reply score_models/analytic_models/conv_likelihood.py]: currently if you want a callable then you use the `f` argument. I bet I can merge them, I'll work on it

**Review by AlexandreAdam — APPROVED** [review]

**Pull request merged**